File: dotnet/SlimLO/Internal/WorkerLocator.cs
using System;
using System.IO;
using System.Runtime.InteropServices;

namespace SlimLO.Internal;

/// <summary>
/// Locates the slimlo_worker executable and resource directory at runtime.
/// </summary>
internal static class WorkerLocator
{
    /// <summary>
    /// Find the slimlo_worker executable.
    /// </summary>
    public static string FindWorkerExecutable()
    {
        var workerName = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? "slimlo_worker.exe" : "slimlo_worker";
        var assemblyDir = GetAssemblyDirectory();

        // Search locations in priority order
        string[] searchPaths = new[]
        {
            // Same directory as assembly (most common for dev/test)
            Path.Combine(assemblyDir, workerName),
            // runtimes/{rid}/native/ (NuGet package layout)
            Path.Combine(assemblyDir, "runtimes", GetRuntimeIdentifier(), "native", workerName),
            // native/ subdirectory
            Path.Combine(assemblyDir, "native", workerName),
            // program/ subdirectory (direct artifact layout)
            Path.Combine(assemblyDir, "program", workerName),
        };

        foreach (var path in searchPaths)
        {
            if (File.Exists(path))
                return Path.GetFullPath(path);
        }

        // Check SLIMLO_WORKER_PATH environment variable
        var envPath = Environment.GetEnvironmentVariable("SLIMLO_WORKER_PATH");
        if (!string.IsNullOrEmpty(envPath) && File.Exists(envPath))
            return Path.GetFullPath(envPath);

        throw new FileNotFoundException(
            $"Cannot find slimlo_worker executable. Searched: {string.Join(", ", searchPaths)}. " +
            "Set SLIMLO_WORKER_PATH environment variable or install the SlimLO.Native NuGet package.");
    }

    /// <summary>
    /// Find the SlimLO resource directory (containing program/, share/).
    /// </summary>
    public static string FindResourcePath()
    {
        var assemblyDir = GetAssemblyDirectory();

        // Search locations in priority order
        string[] candidates = new[]
        {
            // Same directory as assembly has program/ subdirectory
            assemblyDir,
            // slimlo-resources/ subdirectory
            Path.Combine(assemblyDir, "slimlo-resources"),
            // Parent directory
            Path.Combine(assemblyDir, ".."),
            // Parent's slimlo-resources
            Path.Combine(assemblyDir, "..", "slimlo-resources"),
            // NuGet package layout
            Path.Combine(assemblyDir, "runtimes", GetRuntimeIdentifier(), "native"),
        };

        foreach (var candidate in candidates)
        {
            // Resource dir must contain program/ with libmergedlo or mergedlo
            var programDir = Path.Combine(candidate, "program");
            if (Directory.Exists(programDir) && HasMergedLibrary(programDir))
                return Path.GetFullPath(candidate);
        }

        // Check SLIMLO_RESOURCE_PATH environment variable
        var envPath = Environment.GetEnvironmentVariable("SLIMLO_RESOURCE_PATH");
        if (!string.IsNullOrEmpty(envPath) && Directory.Exists(Path.Combine(envPath, "program")))
            return Path.GetFullPath(envPath);

        throw new InvalidOperationException(
            "Cannot auto-detect SlimLO resource path. " +
            "Set SLIMLO_RESOURCE_PATH environment variable or pass ResourcePath in PdfConverterOptions.");
    }

    private static string GetRuntimeIdentifier()
    {
#if NET5_0_OR_GREATER
        return RuntimeInformation.RuntimeIdentifier;
#else
        string os;
        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
            os = "win";
        else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
            os = "osx";
        else
            os = "linux";

        string arch = RuntimeInformation.ProcessArchitecture switch
        {
            Architecture.X64 => "x64",
            Architecture.Arm64 => "arm64",
            Architecture.X86 => "x86",
            Architecture.Arm => "arm",
            _ => "x64"
        };

        return $"{os}-{arch}";
#endif
    }

    private static bool HasMergedLibrary(string programDir)
    {
        // Check for any platform's merged library
        return File.Exists(Path.Combine(programDir, "libmergedlo.so"))
            || File.Exists(Path.Combine(programDir, "libmergedlo.dylib"))
            || File.Exists(Path.Combine(programDir, "mergedlo.dll"))
            || File.Exists(Path.Combine(programDir, "sofficerc"));
    }

    private static string GetAssemblyDirectory()
    {
        var location = typeof(WorkerLocator).Assembly.Location;
        if (string.IsNullOrEmpty(location))
        {
            // Single-file publish: use AppContext.BaseDirectory
            return AppContext.BaseDirectory;
        }
        return Path.GetDirectoryName(location) ?? AppContext.BaseDirectory;
    }
}

File: java/slimlo/src/main/java/com/slimlo/internal/WorkerLocator.java
package com.slimlo.internal;

import java.io.File;
import java.io.FileNotFoundException;
import java.net.URISyntaxException;
import java.net.URL;
import java.security.CodeSource;
import java.security.ProtectionDomain;

/**
 * Locates the slimlo_worker executable and resource directory at runtime.
 */
public final class WorkerLocator {

    private WorkerLocator() {}

    /**
     * Find the slimlo_worker executable.
     *
     * @return Absolute path to the worker executable.
     * @throws FileNotFoundException if the worker cannot be found.
     */
    public static String findWorkerExecutable() throws FileNotFoundException {
        String workerName = isWindows() ? "slimlo_worker.exe" : "slimlo_worker";
        String assemblyDir = getAssemblyDirectory();

        String[] searchPaths = {
                // Same directory as JAR
                new File(assemblyDir, workerName).getAbsolutePath(),
                // native/ subdirectory
                new File(new File(assemblyDir, "native"), workerName).getAbsolutePath(),
                // program/ subdirectory
                new File(new File(assemblyDir, "program"), workerName).getAbsolutePath(),
        };

        for (String path : searchPaths) {
            File f = new File(path);
            if (f.isFile() && f.canExecute()) {
                return f.getAbsolutePath();
            }
        }

        // Check SLIMLO_WORKER_PATH environment variable
        String envWorkerPath = System.getenv("SLIMLO_WORKER_PATH");
        if (envWorkerPath != null && !envWorkerPath.isEmpty()) {
            File f = new File(envWorkerPath);
            if (f.isFile()) {
                return f.getAbsolutePath();
            }
        }

        // Check inside SLIMLO_RESOURCE_PATH/program/
        String envResourcePath = System.getenv("SLIMLO_RESOURCE_PATH");
        if (envResourcePath != null && !envResourcePath.isEmpty()) {
            File f = new File(new File(envResourcePath, "program"), workerName);
            if (f.isFile()) {
                return f.getAbsolutePath();
            }
        }

        StringBuilder searched = new StringBuilder();
        for (String path : searchPaths) {
            if (searched.length() > 0) searched.append(", ");
            searched.append(path);
        }

        throw new FileNotFoundException(
                "Cannot find slimlo_worker executable. Searched: " + searched +
                ". Set SLIMLO_WORKER_PATH or SLIMLO_RESOURCE_PATH environment variable, " +
                "or ensure the native assets are in the expected location.");
    }

    /**
     * Find the SlimLO resource directory (containing program/, share/).
     *
     * @return Absolute path to the resource directory.
     * @throws IllegalStateException if the resource path cannot be found.
     */
    public static String findResourcePath() {
        String assemblyDir = getAssemblyDirectory();

        String[] candidates = {
                // Same directory as JAR
                assemblyDir,
                // slimlo-resources/ subdirectory
                new File(assemblyDir, "slimlo-resources").getAbsolutePath(),
                // Parent directory
                new File(assemblyDir, "..").getAbsolutePath(),
                // Parent's slimlo-resources
                new File(new File(assemblyDir, ".."), "slimlo-resources").getAbsolutePath(),
        };

        for (String candidate : candidates) {
            File programDir = new File(candidate, "program");
            if (programDir.isDirectory() && hasMergedLibrary(programDir)) {
                try {
                    return new File(candidate).getCanonicalPath();
                } catch (Exception e) {
                    return new File(candidate).getAbsolutePath();
                }
            }
        }

        // Check SLIMLO_RESOURCE_PATH environment variable
        String envPath = System.getenv("SLIMLO_RESOURCE_PATH");
        if (envPath != null && !envPath.isEmpty()) {
            File programDir = new File(envPath, "program");
            if (programDir.isDirectory()) {
                return new File(envPath).getAbsolutePath();
            }
        }

        throw new IllegalStateException(
                "Cannot auto-detect SlimLO resource path. " +
                "Set SLIMLO_RESOURCE_PATH environment variable or pass resourcePath " +
                "in PdfConverterOptions.");
    }

    private static boolean hasMergedLibrary(File programDir) {
        return new File(programDir, "libmergedlo.so").exists()
                || new File(programDir, "libmergedlo.dylib").exists()
                || new File(programDir, "mergedlo.dll").exists()
                || new File(programDir, "sofficerc").exists();
    }

    /**
     * Get the directory containing this JAR/class file.
     */
    static String getAssemblyDirectory() {
        try {
            ProtectionDomain pd = WorkerLocator.class.getProtectionDomain();
            if (pd != null) {
                CodeSource cs = pd.getCodeSource();
                if (cs != null) {
                    URL location = cs.getLocation();
                    if (location != null) {
                        File file = new File(location.toURI());
                        // If this is a JAR, return its parent directory
                        if (file.isFile()) {
                            return file.getParentFile().getAbsolutePath();
                        }
                        // If this is a directory (running from classes/)
                        return file.getAbsolutePath();
                    }
                }
            }
        } catch (URISyntaxException e) {
            // Fall through
        }

        // Fallback: current working directory
        return System.getProperty("user.dir", ".");
    }

    static boolean isWindows() {
        String os = System.getProperty("os.name", "").toLowerCase();
        return os.contains("win");
    }

    static boolean isMacOS() {
        String os = System.getProperty("os.name", "").toLowerCase();
        return os.contains("mac") || os.contains("darwin");
    }

    static boolean isLinux() {
        String os = System.getProperty("os.name", "").toLowerCase();
        return os.contains("linux");
    }
}
